code/__HELPERS/mobs.dm

```diff
@@ -346,62 +346,87 @@ This is always put in the attack log.
+ *
+ *	Example usage:
+ *
+ *	if(do_after(user, 50, target = sometarget, extra_checks = list(callback_check1, callback_check2)))
+ *		do_stuff()
+ *	
+ *	This will create progress bar that lasts for 5 seconds. If the user doesn't move or otherwise do something that would cause the checks to fail in those 5 seconds, do_stuff() would execute.
+ *	The Proc returns TRUE upon success (the progress bar reached the end), or FALSE upon failure (the user moved or some other check failed)
+ */
+/proc/do_after(mob/user, delay, needhand = 1, atom/target = null, progress = 1, list/extra_checks = list(), use_default_checks = TRUE)
 	if(!user)
-		return 0
+		return FALSE
 	var/atom/Tloc = null
 	if(target)
 		Tloc = target.loc
 
 	var/atom/Uloc = user.loc
 
-	var/drifting = 0
+	var/drifting = FALSE
 	if(!user.Process_Spacemove(0) && user.inertia_dir)
-		drifting = 1
+		drifting = TRUE
 
 	var/holding = user.get_active_hand()
 
-	var/holdingnull = 1 //User's hand started out empty, check for an empty hand
+	var/holdingnull = TRUE //User's hand started out empty, check for an empty hand
 	if(holding)
-		holdingnull = 0 //Users hand started holding something, check to see if it's still holding that
+		holdingnull = FALSE //Users hand started holding something, check to see if it's still holding that
 
 	var/datum/progressbar/progbar
 	if(progress)
 		progbar = new(user, delay, target)
 
 	var/endtime = world.time + delay
 	var/starttime = world.time
-	. = 1
+	. = TRUE
+
+	// By default, checks for weakness and stunned get added to the extra_checks list.
+	// Setting `use_default_checks` to FALSE means that you don't want the do_after to check for these statuses, or that you will be supplying your own checks.
+	if(use_default_checks)
+		extra_checks += CALLBACK(user, /mob.proc/IsWeakened)
+		extra_checks += CALLBACK(user, /mob.proc/IsStunned)
+
 	while(world.time < endtime)
 		sleep(1)
 		if(progress)
 			progbar.update(world.time - starttime)
 
 		if(drifting && !user.inertia_dir)
-			drifting = 0
+			drifting = FALSE
 			Uloc = user.loc
 
-		if(!user || user.stat || user.IsWeakened() || user.stunned  || (!drifting && user.loc != Uloc)|| (extra_checks && !extra_checks.Invoke()))
-			. = 0
+		if(!user || user.stat || (!drifting && user.loc != Uloc) || invoke_callback_list(extra_checks))
+			. = FALSE
 			break
 
 		if(Tloc && (!target || Tloc != target.loc))
-			. = 0
+			. = FALSE
 			break
 
 		if(needhand)
 			//This might seem like an odd check, but you can still need a hand even when it's empty
 			//i.e the hand is used to pull some item/tool out of the construction
 			if(!holdingnull)
 				if(!holding)
-					. = 0
+					. = FALSE
 					break
 			if(user.get_active_hand() != holding)
-				. = 0
+				. = FALSE
 				break
 	if(progress)
 		qdel(progbar)
 
+// Upon any of the callbacks in the list returning TRUE, the proc will return TRUE.
+/proc/invoke_callback_list(list/extra_checks)

```

Comment:
I suggest renaming this to `any_callback_true` or similar, because right now the semantics are unobvious, if you see this called in the code somewhere and don't bother checking the comment in the definition. I originally thought it would return true if *all* the callbacks return true, but this is not the case here.

Reply:
That's a funky sounding proc name, but I guess its name is more telling of its purpose than my current name.